Delete old images of azureml tests periodically

miguelgfierro opened the pull request:
### Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
This explains how to delete old images stored in container registry during the test process.

### Related Issues
<!--- If it fixes an open issue, please link to the issue here. -->


### References
<!--- References would be helpful to understand the changes. -->
<!--- References can be books, links, etc. -->


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have followed the [contribution guidelines](CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [ ] I have [signed the commits](https://github.com/recommenders-team/recommenders/wiki/How-to-sign-commits), e.g. `git commit -s -m "your commit message"`. 
- [ ] This PR is being made to `staging branch` AND NOT TO `main branch`.

miguelgfierro commented:
to schedule the task: https://learn.microsoft.com/en-us/azure/container-registry/container-registry-auto-purge#run-in-a-scheduled-task

miguelgfierro commented on tests/README.md:
```diff
@@ -332,3 +333,28 @@ Example:
     def test_to_skip():
         assert False
 
+## Other relevant information
```
ok, with the commands above I can erase the images but not the repositories. I see two options:
1. Identify empty repositories and erase them:
```
az acr repository list --name 978a92daa2ad4447aae1b21196dd4a9b --query "[?starts_with(@, 'azureml/')]" -o tsv | while read repo; do
    tags=$(az acr repository show-tags --name 978a92daa2ad4447aae1b21196dd4a9b --repository "$repo" --query "[]" -o tsv)
    if [ -z "$tags" ]; then
        echo "Deleting empty repository: $repo"
        az acr repository delete --name 978a92daa2ad4447aae1b21196dd4a9b --repository "$repo" --yes
    fi
done
```
2. Instead of naming the each repository and image differently in every azureml run, we can have one repository and different images. This might be easier for automating the periodic deleting. @SimonYansenZhao do you know when in the test pipeline are we naming the repositories so we can fix them?

SimonYansenZhao replied:
The only code for image creation is this: https://github.com/recommenders-team/recommenders/blob/45e1b215a35e69b92390e16eb818d4528d0a33a2/tests/ci/azureml_tests/aml_utils.py#L101-L110

But the names of the images (such as `recommenders-xxx-python3_11-spark`) are meaningful: https://github.com/recommenders-team/recommenders/blob/45e1b215a35e69b92390e16eb818d4528d0a33a2/.github/workflows/azureml-unit-tests.yml#L73
They aren't like those (such as `978a92daa2ad4447aae1b21196dd4a9b `) in your example.

I have no idea of how to control the naming of the repositories.  And I don't find there is any argument of [the Environment class](https://learn.microsoft.com/en-us/python/api/azure-ai-ml/azure.ai.ml.entities.environment?view=azure-python) in the AzureML SDK can be used for that.

SimonYansenZhao replied:
@miguelgfierro Will appending the creation timestamp of the environment to its name be helpful for you to find the old images for deletion?  I mean, changing the name of the environment such as `recommenders-xxx-python3_11-spark` to `recommenders-xxx-python3_11-spark-2025-05-17-14-43` will be helpful?

miguelgfierro replied:
example of a repository and image

![image](https://github.com/user-attachments/assets/0002cbb9-4c71-44f1-9c3a-a86b66905216)

miguelgfierro replied:
Original size:
![image](https://github.com/user-attachments/assets/3a4a9d5a-25ad-4a91-9107-9d7533cd0363)

miguelgfierro commented:
Iterating:
]$ az acr task create --name delete-empty-repos --registry 978a92daa2ad4447aae1b21196dd4a9b --cmd 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest --context https://github.com/recommenders-team/acr-script.git --schedule "43 12 * * *" --timeout 3600 --commit-trigger-enabled false --pull-request-trigger-enabled false --platform linux
missing --file/-f argument

The error missing --file/-f argument occurs because the --context parameter in the az acr task create command is specified (pointing to your Git repository https://github.com/recommenders-team/acr-script.git), but the --file argument, which specifies the Dockerfile or task YAML file, is missing. When using a Git context with --context, the Azure CLI expects a --file to indicate the task definition file (e.g., a Dockerfile or a YAML file defining task steps) unless the task only runs a pre-built image without building.
Since your goal is to run the pre-built image 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest (which executes the delete_empty_repos.sh script via the CMD in your Dockerfile) and schedule it daily at 12:43 UTC (from --schedule "43 12 * * *"), you can fix the issue by:
Removing the --context since you’re not building an image during the task run (the dummy:latest image is already built and pushed).

Using --context /dev/null to explicitly indicate no build context is needed, similar to your working purge_images_15dago task.

Keeping the --cmd to run the pre-built image, which executes the script.

Updated One-Liner
bash

az acr task create --name delete-empty-repos --registry 978a92daa2ad4447aae1b21196dd4a9b --cmd 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest --context /dev/null --schedule "43 12 * * *" --timeout 3600 --commit-trigger-enabled false --pull-request-trigger-enabled false --platform linux

Breakdown
--name delete-empty-repos: Task name.

--registry 978a92daa2ad4447aae1b21196dd4a9b: Your ACR.

--cmd 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest: Runs the pre-built image, which executes CMD ["/bin/bash", "/delete_empty_repos.sh"] from your Dockerfile.

--context /dev/null: Specifies no build context, as the task only runs the image, not builds it (matches your purge_images_15dago task).

--schedule "43 12 * * *": Runs daily at 12:43 UTC.

--timeout 3600: 1-hour timeout for large registries.

--commit-trigger-enabled false --pull-request-trigger-enabled false: Disables Git triggers for schedule-based execution.

--platform linux: Ensures a Linux container.

Why the Error Occurred
The original command included --context https://github.com/recommenders-team/acr-script.git without --file. ACR Tasks interpret this as an attempt to build an image from the Git repository, requiring a --file (e.g., Dockerfile or a YAML task file). Since your task only needs to run the pre-built dummy:latest image, --context should be /dev/null to skip the build step.

Steps to Apply
Run the Updated Command:
If the task doesn’t exist, create it:
bash

az acr task create --name delete-empty-repos --registry 978a92daa2ad4447aae1b21196dd4a9b --cmd 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest --context /dev/null --schedule "43 12 * * *" --timeout 3600 --commit-trigger-enabled false --pull-request-trigger-enabled false --platform linux

If the task already exists, update it:
bash

az acr task update --name delete-empty-repos --registry 978a92daa2ad4447aae1b21196dd4a9b --cmd 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest --context /dev/null --schedule "43 12 * * *" --timeout 3600 --commit-trigger-enabled false --pull-request-trigger-enabled false --platform linux

Ensure the Image Exists:
Verify that 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest exists in your ACR:
bash

az acr repository show-tags --name 978a92daa2ad4447aae1b21196dd4a9b --repository dummy --output table

If it’s missing, rebuild and push it using your Dockerfile:
bash

docker build -t 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest -f Dockerfile .
az acr login --name 978a92daa2ad4447aae1b21196dd4a9b
docker push 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest

Test the Task:
Run the task manually to confirm the script executes:
bash

az acr task run --name delete-empty-repos --registry 978a92daa2ad4447aae1b21196dd4a9b

Check Logs:
Verify the script’s output (e.g., Deleted empty repository: <repo>):
bash

az acr task show-runs --name delete-empty-repos --registry 978a92daa2ad4447aae1b21196dd4a9b --output table
az acr task logs --name delete-empty-repos --registry 978a92daa2ad4447aae1b21196dd4a9b --run-id <run-id>

Verify Repository Deletion:
Check remaining repositories:
bash

az acr repository list --name 978a92daa2ad4447aae1b21196dd4a9b --output table

Notes
Authentication: Your environment is already authenticated (as confirmed by the purge_images_15dago task). The mcr.microsoft.com/azure-cli image in dummy:latest uses these credentials for az acr commands. If you encounter authentication errors, ensure the AcrDelete role is assigned:
bash

az role assignment create --assignee <principal-id> --role AcrDelete --scope /subscriptions/<subscription-id>/resourceGroups/<resource-group>/providers/Microsoft.ContainerRegistry/registries/978a92daa2ad4447aae1b21196dd4a9b

Manifests: If repositories aren’t deleted due to untagged manifests, update delete_empty_repos.sh to check manifests:
bash

#!/bin/bash
registry="978a92daa2ad4447aae1b21196dd4a9b"

az acr repository list --name $registry --output tsv | while read repo; do
  if [ -z "$(az acr repository show-tags --name $registry --repository $repo --output tsv 2>/dev/null)" ] && [ -z "$(az acr repository show-manifests --name $registry --repository $repo --output tsv 2>/dev/null)" ]; then
    az acr repository delete --name $registry --repository $repo --yes
    echo "Deleted empty repository: $repo"
  else
    echo "Repository $repo is not empty, skipping..."
  fi
done

Update the Git repository (recommenders-team/acr-script), rebuild dummy:latest, and push it:
bash

docker build -t 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest -f Dockerfile .
az acr login --name 978a92daa2ad4447aae1b21196dd4a9b
docker push 978a92daa2ad4447aae1b21196dd4a9b.azurecr.io/dummy:latest

Timeout: The 3600-second timeout should suffice. Increase to 7200 seconds if needed:
bash

--timeout 7200

Schedule: The schedule 43 12 * * * runs at 12:43 UTC daily. Adjust if needed (e.g., 0 12 * * * for 12:00 UTC).

Summary
Issue: The original task failed due to a missing --file with --context. Since you’re running a pre-built image, --context /dev/null is appropriate.

Fix: Updated the one-liner to use --cmd with dummy:latest and --context /dev/null, matching your purge_images_15dago task’s structure.

Next Steps: Run the updated create or update command, test manually, and check logs for script execution.

If you see errors (e.g., authentication or script execution issues), share the detailed logs (az acr task logs), and I’ll assist with troubleshooting!

miguelgfierro commented:
After discussing with Simon, we are going to try to add something like this in the `post_pytest.py` file:
```
from azure.containerregistry import ContainerRegistryClient
from azure.identity import DefaultAzureCredential

# Replace with your ACR name (e.g., from Azure ML workspace: ws.container_registry.address.split('.')[0])
acr_name = "youracrname"
endpoint = f"https://{acr_name}.azurecr.io"
credential = DefaultAzureCredential()

client = ContainerRegistryClient(endpoint=endpoint, credential=credential)

# Example: Delete a specific tag from a repository
repository_name = "your-repository"  # e.g., 'samples/python-app'
tag = "latest"  # or use manifest digest like 'sha256:abc123...'
client.delete_tag(repository_name, tag)

# Alternatively, delete an entire repository (all images and tags)
# client.delete_repository(repository_name)

client.close()
```

miguelgfierro commented:
I think I was able to put it to work @SimonYansenZhao.

The latest command is scheduled to delete everything, both images and repositories. 

I'll wait for tomorrow to make sure that it works, and if there are no errors, I will merge this.
